lib/shelley/src/Cardano/Wallet/Shelley/Network/Node.hs

```diff
@@ -1043,70 +1001,52 @@ recoveringNodeConnection tr handlers action =
-    -- In such case, we simply retry a bit later and hope it's there.
-    | isDoesNotExistError e =
-        pure ConsultPolicy
-
-    -- If the nonblocking UNIX domain socket connection cannot be completed
-    -- immediately (i.e. connect() returns EAGAIN), try again. This happens
-    -- because the node's listen queue is quite short.
-    | isTryAgainError e =
-        pure ConsultPolicy
-
-    | isResourceVanishedError e = do
-        traceWith tr $ Just e
-        pure $ if onResourceVanished then ConsultPolicy else DontRetry
-
-    | otherwise = do
-        pure DontRetry
-  where
-    isResourceVanishedError = isInfixOf "resource vanished" . show
-    isTryAgainError = isInfixOf "resource exhausted" . show
-
-handleMuxError
-    :: Tracer IO (Maybe IOException)
-    -> Bool -- ^ 'True' = retry on 'ResourceVanishedError'
-    -> MuxError
-    -> IO RetryAction
-handleMuxError tr onResourceVanished = pure . errorType >=> \case
-    MuxUnknownMiniProtocol -> pure DontRetry
-    MuxDecodeError -> pure DontRetry
-    MuxIngressQueueOverRun -> pure DontRetry
-    MuxInitiatorOnly -> pure DontRetry
-    MuxShutdown _ -> pure DontRetry -- fixme: #2212 consider cases
-    MuxCleanShutdown -> pure DontRetry
-    MuxIOException e ->
-        handleIOException tr onResourceVanished e
-    MuxBearerClosed -> do
-        traceWith tr Nothing
-        pure $ if onResourceVanished then ConsultPolicy else DontRetry
-
-    -- MuxSDU*Timeout errors arise because the bandwidth of the
-    -- interprocess communication socket dropped unexpectedly,
-    -- and the socket library decided to cut off the connection
-    -- after ~ 30 seconds.
-    -- Chances are that the system is overloaded, let's retry in 30 seconds.
-    MuxSDUReadTimeout -> pure $ ConsultPolicyOverrideDelay 30_000_000
-    MuxSDUWriteTimeout -> pure $ ConsultPolicyOverrideDelay 30_000_000
+handleIOException :: Tracer IO IOException -> IOException -> IO RetryAction
+handleIOException tr ioException =
+    traceWith tr ioException $> case ioException of
+        -- There's a race-condition when starting the wallet and the node at the
+        -- same time: the socket might not be there yet when we try to open it.
+        -- In such case, we simply retry a bit later and hope it's there.
+        _ | isDoesNotExistError ioException -> ConsultPolicy
+        -- I/O error where the operation failed because the resource vanished.
+        -- This happens when, for example, attempting to write to a closed
+        -- socket or attempting to write to a named pipe that was deleted.
+        e | isResourceVanishedError e -> ConsultPolicy
+        -- If the nonblocking UNIX domain socket connection cannot be completed
+        -- immediately (i.e. connect() returns EAGAIN), try again. This happens
+        -- because the node's listen queue is quite short.
+        e | "resource exhausted" `isInfixOf` show e -> ConsultPolicy
+        _ -> DontRetry
```

Comment:
I think that this case statement is harder to understand than the previous formulation in terms of guards — the name `e` is brought into scope, but it's always equal to `ioException`. If you really want to use guards inside an expression, I recommend `MultiWayIf`, otherwise I'd just use the previous formulation.

Reply:
🆗